libnmstate/nm/wireguard.py

```diff
@@ -0,0 +1,145 @@
+#
+# Copyright (c) 2021 Red Hat, Inc.
+#
+# This file is part of nmstate
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+from libnmstate.schema import Wireguard
+
+from .common import NM
+from .utils import get_secrets
+
+
+def create_wireguard_setting(wireguard_iface, base_con_profile):
+    wireguard_setting = None
+    if base_con_profile:
+        wireguard_setting = base_con_profile.get_setting_by_name(
+            NM.SETTING_WIREGUARD_SETTING_NAME
+        )
+        if wireguard_setting:
+            wireguard_setting = wireguard_setting.duplicate()
+
+    if not wireguard_setting:
+        wireguard_setting = NM.SettingWireGuard.new()
+    if wireguard_iface.fwmark:
+        wireguard_setting.props.fwmark = wireguard_iface.fwmark
+    if wireguard_iface.listen_port:
+        wireguard_setting.props.listen_port = wireguard_iface.listen_port
+    if wireguard_iface.private_key:
+        wireguard_setting.props.private_key = wireguard_iface.private_key
+
+    _set_wireguard_peers(wireguard_setting, wireguard_iface.peers)
+
+    return wireguard_setting
+
+
+def _set_wireguard_peers(wireguard_setting, peers_config):
+    wireguard_setting.clear_peers()
+    for peer in peers_config:
+        wireguard_peer = NM.WireGuardPeer.new()
+        wireguard_peer.set_endpoint(peer.get(Wireguard.Peer.ENDPOINT), True)
+        if peer.get(Wireguard.Peer.PERSISTENT_KEEPALIVE):
+            wireguard_peer.set_persistent_keepalive(
+                peer[Wireguard.Peer.PERSISTENT_KEEPALIVE]
+            )
+
+        if peer.get(Wireguard.Peer.PRESHARED_KEY):
+            wireguard_peer.set_preshared_key_flags(0)
+            wireguard_peer.set_preshared_key(
+                peer.get(Wireguard.Peer.PRESHARED_KEY), True
+            )
+        if peer.get(Wireguard.Peer.PUBLIC_KEY):
+            wireguard_peer.set_public_key(
+                peer.get(Wireguard.Peer.PUBLIC_KEY), True
+            )
+
+        for allowed_ip in peer.get(Wireguard.Peer.ALLOWED_IPS, []):
+            wireguard_peer.append_allowed_ip(allowed_ip, True)
+
+        wireguard_setting.append_peer(wireguard_peer)
+
+
+def get_info(context, nm_ac):
+    if not nm_ac:
+        return {}
+    nm_profile = nm_ac.get_connection()
```

Comment:
Even this is for temporary use(we will move to Nispor for querying the kernel interface), but please use `applied_config` if possible. Please check `nm/ipv4.py` for example.

Reply:
Hi Gris, this won't work. applied_config is a `NM.SimpleConnection` and that class does not contain `get_secrets_async()` that is needed to get the secrets of Wireguard. The class required is `NM.RemoteConnection` which is the one that `nm_ac.get_connection()` is provided. I've been working with Shreya on this but I think unfortunately we cannot use `applied_config` here.